### PR TITLE
[dtensor][random][tp] remove the adhoc DTensor RNG tracker TensorParallelRNGTracker since it does not match FSDP2+TP

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #141532
* #141223
* __->__ #141220
* #141731

**Summary**
The ad-hoc DTensor RNG tracker was used to mimic Megatron DDP+TP RNG behavior but it turns out not compatible with PyTorch Distributed FSDP2+TP so we decide to deprecate it and use `OffsetBasedRNGTracker` to replace, which follows the SPMD semantics (replicas get the same random sampling result, shards get different results).

**Motivation**
`TensorParallelRNGTracker` was designed for DDP+TP where the random operators produce the same result along the data parallel mesh dimension and different results along the tensor parallel dimension. However this does not apply to the new FSDP+TP composable combination where the model weights are sharded along data parallel mesh dimension as well. Therefore we decide to remove this outdated RNG tracker type for now. If users have demands for exact match between PyTorch Distributed and Megatron on Random Number generation result, feel free to file an issue.

**Impact**
`TensorParallelRNGTracker` was only used when Tensor Parallel is used (i.e. calling `parallelize_module`). 

For non-FSDP users, the "replicas get the same random numbers and shards get different ones" remains unchanged. Unlike `TensorParallelRNGTracker` which sets different seeds (`base_seed + 2718 + TP_rank`) within the TP group, DTensor now sets the same seed (default value is 1234 but users can call `torch.distributed.tensor._random.manual_seed` to modify) on all ranks but choose the right RNG offset based on DTensor placements to enforce the "replicas get the same random numbers and shards get different ones" invariant. 

For FSDP2 users, improvement should be observed in a way that DTensor sharded within DP group now gets different random number sampling which `TensorParallelRNGTracker` failed to do, though we're not sure how much this change will improve the eventual training loss convergence.

**Test**
1-d model weight meta init:
`pytest test/distributed/_tensor/test_random_ops.py -s -k test_tp_model_meta_init`

2-d model weight meta init:
`pytest test/distributed/_tensor/test_random_ops.py -s -k test_fsdp_tp_model_meta_init`

TP model weight init test:
`pytest test/distributed/tensor/parallel/test_tp_random_state.py`

FSDP+TP model weight init test:
`pytest test/distributed/_composable/fsdp/test_fully_shard_init.py`

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @tianyu-l

### CODE REVIEW

**Comment by wconstab** [comment]
I think the change looks good, but we should offer more thorough documentation of the reasons and consequences of the change in the PR desc.
1. why TensorParallelRNGTracker was wrong?
2. in what situations will existing users be impacted? in what situations will existing users show no change
3. did we do any test to ensure OffsetBasedRNGTracker actually works as expected for initializing a 1D, 2D model?

File: torch/distributed/tensor/_random.py [hunk 1]
```diff
@@ -89,9 +88,7 @@ def manual_seed(seed: int, device_mesh: DeviceMesh) -> None:
 
     # the current rank is in mesh
     if device_mesh.get_coordinate() is not None:
-        if isinstance(_rng_tracker, TensorParallelRNGTracker):
-            _rng_tracker._manual_seed(device_mesh, seed)
-        elif isinstance(_rng_tracker, OffsetBasedRNGTracker):
+        if isinstance(_rng_tracker, OffsetBasedRNGTracker):
```
Comment (tianyu-l) [inline torch/distributed/tensor/_random.py:91]: nit: if we only keep one type of rng_tracker, we don't need the if-else here

File: torch/distributed/tensor/_random.py [hunk 2]
```diff
@@ -260,23 +255,48 @@ def _set_pre_op_offset(self, spec: DTensorSpec) -> None:
         """
         dtensor_shape = spec.shape
         mesh = spec.mesh
-        dim_map = spec.dim_map
+        # note: dim_map does not allow double sharding which is the FSDP(fully_shard)+TP
+        # case. Replace the custom logic with dim_map once we support it.
+        dim_map: List[Union[int, List[int]]] = [-1] * spec.ndim
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:260]: so dim_map will contain a mapping of 
"mesh dim" -> "tensor dims that are sharded over this mesh dim"
?

**Comment by wconstab** [comment]
> Test
TP model weight init test: pytest test/distributed/tensor/parallel/test_tp_random_state.py
FSDP+TP model weight init test: pytest test/distributed/_composable/fsdp/test_fully_shard_init.py

Can you say more about the test plan?  I expect we need new tests or updates to existing tests.  If existing tests were passing without this PR landing, then it proves they do not cover the case we need coverage for.  Particularly, I would expect the test to use meta-init for a model that has tensor-parallel and fully_shard applied to it, move to empty, then initialize.  We should check that each shard gets unique values.  (and this should fail if we use the old TensorParallelRNGTracker)

File: torch/distributed/tensor/_random.py (hunk 2 above)
Comment (XilunWu) [reply torch/distributed/tensor/_random.py:260]: the opposite. `dim_map` is "tensor dim" -> "mesh dims that the tensor dim is sharded over".

File: torch/distributed/tensor/_random.py [hunk 3]
```diff
@@ -260,23 +255,48 @@ def _set_pre_op_offset(self, spec: DTensorSpec) -> None:
...
+            if isinstance(mesh_dim, List):
+                assert isinstance(mesh_dim, List)
+                rank_coord = [mesh_coordinate[d] for d in mesh_dim]
+                num_shards = [mesh_size[d] for d in mesh_dim]
+                # compute the shard idx and total number of shards
+                for idx, size in zip(rank_coord, num_shards):
+                    shard_idx = shard_idx * size + idx
+                    total_num_shards *= size
+
+            shard_idx_by_dim.append(shard_idx)
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py:293]: ah, so we make a flattened 'shard_idx' for the cases where we have a tensor dim sharded over multiple mesh dims.  so instead of having 'shard idx = (1,0)' we get 'shard idx=3'

**Review by wconstab — APPROVED** [review]
